[jit] Add support for dynamic shape fusion in JIT.

navahgar opened the pull request:
Summary:
This diff adds support for dynamic shape fusion in JIT. This is done
by performing fusion with the static shapes observed on the first run,
generalizing the fused subgraphs and generating code for the generalized fused
subgraphs with dynamic shapes.

Test Plan:
```
buck test mode/dev-nosan //caffe2/test/cpp/jit:jit
```

Differential Revision: D32781307

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -489,10 +501,17 @@ class TensorExprFuser {
+    if (tensorExprDynamicShapeFusionEnabled()) {
+      VLOG(1) << "TensorExpr fusion with dynamic shapes is enabled"
+              << std::endl;
+      generalizeFusionGroups(graph_->block());
```
add TODO here: removeTensorTypeSpecializations

eellison reviewed (commented):
Nice !! Could we add one test ?

navahgar committed:
[StaticRuntime] Refactor StaticModule to pass in sample inputs.

Summary:
This diff refactors StaticModule and its uses to pass in sample
inputs.

Test Plan: buck run mode/opt //caffe2/caffe2/fb/predictor:pytorch_predictor_test

Differential Revision: D32320084

fbshipit-source-id: d16c9c33dfbedf527bbca9757b0bbe34d8e2c085

navahgar committed:
[jit] Add support for dynamic shape fusion in JIT. (#69269)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/69269

This diff adds support for dynamic shape fusion in JIT. This is done
by performing fusion with the static shapes observed on the first run,
generalizing the fused subgraphs and generating code for the generalized fused
subgraphs with dynamic shapes.

Test Plan:
```
buck test mode/dev-nosan //caffe2/test/cpp/jit:jit
```

Differential Revision: D32781307

fbshipit-source-id: 0a72228e29785785459ab1c14217c600971637a7
